cpp/src/arrow/dataset/scanner.h

```diff
@@ -29,31 +29,73 @@
 namespace arrow {
 namespace dataset {
 
-/// \brief Shared state for a Scan operation
-struct ARROW_DS_EXPORT ScanContext {
-  MemoryPool* pool = arrow::default_memory_pool();
-};
-
-// TODO(wesm): API for handling of post-materialization filters. For
-// example, if the user requests [$col1 > 0, $col2 > 0] and $col1 is a
-// partition key, but $col2 is not, then the filter "$col2 > 0" must
-// be evaluated in-memory against the RecordBatch objects resulting
-// from the Scan
-
-class ARROW_DS_EXPORT ScanOptions {
+/// Container for scan state
+struct ARROW_DS_EXPORT ScanOptions final {
  public:
-  virtual ~ScanOptions() = default;
+  ScanOptions() = default;
 
+  /// Filters
   const std::shared_ptr<DataSelector>& selector() const { return selector_; }
 
+  ScanOptions& selector(std::shared_ptr<DataSelector> s) {
```

Comment:
Not sure about the naming. Should this kind of method be called `set_selector`? @wesm